scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,325 @@
+// Import built in libraries needed.
+const fs = require('fs');
+
+// Set Genesys Cloud objects
+const platformClient = require('purecloud-platform-client-v2');
+const apiClient = platformClient.ApiClient.instance;
+// API Instances that will be used to fetch external contacts, organizations, notes and relationships from Genesys Cloud
+const contactsApi = new platformClient.ExternalContactsApi();
+
+// OAuth Client Grant Credentials
+const apiClientId = '-- id here --';
+const apiClientSecret = '-- secret here --';
+const apiClientRegion = '-- genesys cloud region here - ex: mypurecloud.com --';
+
+// Script Sample Settings
+// - Maximum number of records we will request per "page" when using the External Contacts Scan API endpoints (10 to 200)
+const maxRecordsPerPage = 100;
+// - Force the script to make a pause between 2 consecutive API requests (to control/prevent API Rate limit)
+// i.e. Minimum time between 2 API requests
+const minAPIRequestInterval = 500;
+// i.e. If there is a pause to make (processing time < minAPIRequestInterval), the pause will be at minimum minWaitTime
+const minWaitTime = 50;
+// - Define a maximum number of retries (for Login/Authorization errors, for API Rate Limit rejects)
+const maxLoginRetryAttempts = 1;
+const maxRateLimitRetryAttempts = 2;
+// - Settings for Export to file (CSV or RAW)
+const fieldSeparator = ',';
+const exportType = 'CSV';
+
+// Global vars
+var writeStream;
+var consecutiveLoginRetryAttempts = 0;
+var consecutiveRateLimitRetryAttempts = 0;
+
+/*
+------------------------------------------------------------------------------
+* Login, Scan Objects
+------------------------------------------------------------------------------
+*/
+
+//#region Login, Scan
+
+// Login using Client Credentials Grant (returns: true (success), false (failure))
+const login = async () => {
+
+    try {
+        apiClient.setEnvironment(apiClientRegion);
+
+        let creds = await apiClient.loginClientCredentialsGrant(apiClientId, apiClientSecret);
+
+        console.log('Login Success. Token will expire at: ', creds.tokenExpiryTime);
+
+        return true;
+    }
+    catch (error) {
+        console.error('Login error - Error:', JSON.stringify(error, null, 2));
+
+        return false;
+    }
+}
+
+// Perform an External Contacts Scan API request and manage errors (401, 429)
+const gcloudContactsScan = async (cursor) => {
```

Comment:
"gcloud" isn't an acceptable shortening of Genesys Cloud; use the full name or avoid it altogether. (per marketing, don't kill the messenger)

Reply:
I wouldnt even include the name in the variable.  contactsScan is pretty descriptive.